plugins/repository-gcs/src/main/java/org/elasticsearch/common/blobstore/gcs/GoogleCloudStorageBlobContainer.java

```diff
@@ -0,0 +1,145 @@
+import com.google.api.client.googleapis.json.GoogleJsonResponseException;
+import org.elasticsearch.common.blobstore.BlobMetaData;
+import org.elasticsearch.common.blobstore.BlobPath;
+import org.elasticsearch.common.blobstore.BlobStoreException;
+import org.elasticsearch.common.blobstore.support.AbstractBlobContainer;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.repositories.gcs.GoogleCloudStorageClient;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.net.HttpURLConnection.HTTP_NOT_FOUND;
+
+
+public class GoogleCloudStorageBlobContainer extends AbstractBlobContainer {
+
+    protected static final ESLogger logger = Loggers.getLogger(GoogleCloudStorageBlobContainer.class);
+
+    private final GoogleCloudStorageClient client;
+    private final String bucket;
+    private final String keyPath;
+
+    GoogleCloudStorageBlobContainer(BlobPath path, String bucket, GoogleCloudStorageClient client) {
+        super(path);
+        this.client = client;
+        this.bucket = bucket;
+
+        String keyPath = path.buildAsString("/");
+        if (!keyPath.isEmpty()) {
+            keyPath = keyPath + "/";
+        }
+        this.keyPath = keyPath;
+    }
+
+    @Override
+    public boolean blobExists(String blobName) {
+        try {
+            return client.blobExists(bucket, buildKey(blobName));
+        } catch (GoogleJsonResponseException e) {
+            GoogleJsonError error = e.getDetails();
+            if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
+                return false;
+            }
+            throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists in bucket [" + bucket + "]", e);
+        } catch (Exception e) {
+            throw new BlobStoreException("Failed to check if blob [" + blobName + "] exists in bucket [" + bucket + "]", e);
+        }
+    }
+
+    @Override
+    public Map<String, BlobMetaData> listBlobs() throws IOException {
+        return client.listBlobs(bucket, keyPath);
+    }
+
+    @Override
+    public Map<String, BlobMetaData> listBlobsByPrefix(String prefix) throws IOException {
+        return client.listBlobsByPrefix(bucket, keyPath, prefix);
+    }
+
+    @Override
+    public InputStream readBlob(String blobName) throws IOException {
+        try {
+            return client.readBlob(bucket, buildKey(blobName));
+        } catch (GoogleJsonResponseException e) {
+            GoogleJsonError error = e.getDetails();
+            if ((e.getStatusCode() == HTTP_NOT_FOUND) || ((error != null) && (error.getCode() == HTTP_NOT_FOUND))) {
```

Comment:
Or maybe move all checks here. It's just having some checks inside client and other inside container is confusion. I have slight preference for moving all checks like this inside the client. On the second thought why do we need client to start with? If we move all checks inside client this becomes very thin facade for the client, which basically passes everything into the client. So, in order to remove complexity, maybe it makes sense to get rid of client (which is also quite thin facade for Google API) and just move all code from client here.